src/display/display_utils.js

```diff
@@ -658,11 +658,21 @@ class OutputScale {
    * @returns {boolean} Returns `true` if scaling was limited,
    *   `false` otherwise.
    */
-  limitCanvas(width, height, maxPixels, maxDim) {
+  limitCanvas(width, height, maxPixels, maxDim, capCanvasAreaFactor) {
     let maxAreaScale = Infinity,
       maxWidthScale = Infinity,
       maxHeightScale = Infinity;
 
+    if (capCanvasAreaFactor >= 0) {
+      const { innerWidth, innerHeight } = window;
+      maxPixels = Math.min(
+        maxPixels,
+        innerWidth *
+          innerHeight *
+          (1 + capCanvasAreaFactor / 100) *
+          OutputScale.pixelRatio ** 2
+      );
```

Comment:
Sorry, but this really won't work if `maxPixels` is either `-1` or `0` both of which are possible/valid values for that parameter (note the condition just below).
```suggestion
      const capPixels = 
        innerWidth *
          innerHeight *
          (1 + capCanvasAreaFactor / 100) *
          OutputScale.pixelRatio ** 2;

      maxPixels = maxPixels > 0 ? Math.min(maxPixels, capPixels) : capPixels;
```